### PR TITLE
✨ Adds failure domain to machine spec

### PR DESCRIPTION
**Pull request by chuckha**
Signed-off-by: Chuck Ha <chuckh@vmware.com>

**What this PR does / why we need it**: Failure domain seems to be missing from MachineSpec as outline in #1647. This PR ads the FailureDomain and regenerates all the generated files.

/cc @vincepri

### CODE REVIEW

File: api/v1alpha2/zz_generated.conversion.go [hunk 1]
```diff
@@ -958,6 +958,7 @@ func autoConvert_v1alpha3_MachineSpec_To_v1alpha2_MachineSpec(in *v1alpha3.Machi
 	out.InfrastructureRef = in.InfrastructureRef
 	out.Version = (*string)(unsafe.Pointer(in.Version))
 	out.ProviderID = (*string)(unsafe.Pointer(in.ProviderID))
+	// WARNING: in.FailureDomain requires manual conversion: does not exist in peer-type
```
Comment (vincepri) [inline api/v1alpha2/zz_generated.conversion.go]: We need to add this to the conversion webhook, making sure we don't lose the field when down-converting and then back up. There is an example and tests in conversion.go files, lemme know if you need more pointers

Comment (chuckha) [reply api/v1alpha2/zz_generated.conversion.go]: ah, no problems.

what's the situation where we down convert then up convert?

Comment (chuckha) [reply api/v1alpha2/zz_generated.conversion.go]: updated!

File: api/v1alpha2/conversion.go [hunk 2]
```diff
@@ -111,6 +111,9 @@ func restoreMachineSpec(restored *v1alpha3.MachineSpec, dst *v1alpha3.MachineSpe
 		dst.ClusterName = restored.ClusterName
 	}
 	dst.Bootstrap.DataSecretName = restored.Bootstrap.DataSecretName
+	if restored.FailureDomain != nil {
+		dst.FailureDomain = restored.FailureDomain
+	}
```
Comment (vincepri) [inline api/v1alpha2/conversion.go]: This can probably just be `dst.FailureDomain = restored.FailureDomain`

Comment (chuckha) [reply api/v1alpha2/conversion.go]: ack, fixed

**Comment by chuckha** [comment]
@vincepri What situations go from down convert to up convert?

**Comment by vincepri** [comment]
Let's say you have a client that only speaks v1a2, and the stored version is v1a3. The conversion process will get the stored version, downconvert to v1a2, do computation, convert back up to v1a3, store.

**Comment by chuckha** [comment]
> Let's say you have a client that only speaks v1a2, and the stored version is v1a3. The conversion process will get the stored version, downconvert to v1a2, do computation, convert back up to v1a3, store.

ah yeah, got it, thanks for explaining. makes perfect sense 👍

**Comment by chuckha** [comment]
/hold

this actually seems wrong, the spec should have some way to identify the failure domain

**Comment by vincepri** [comment]
Ah @chuckha, I remember now that I actually thought about just adding the FailureDomainID, instead of the spec

**Comment by chuckha** [comment]
> Ah @chuckha, I remember now that I actually thought about just adding the FailureDomainID, instead of the spec

ah ha! I was thinking about adding a `name` field to the FailureDomainSpec, but using ID is fine with me too. The machine controller would have to look up information from the cluster in that case though (which is probably fine)

**Comment by vincepri** [comment]
Do you mean the infrastructure machine controller? In that case, it should be ok, given that the information for failure domains comes from the same provider

File: api/v1alpha3/machine_types.go [hunk 3]
```diff
@@ -72,6 +72,11 @@ type MachineSpec struct {
 	// be interfacing with cluster-api as generic provider.
 	// +optional
 	ProviderID *string `json:"providerID,omitempty"`
+
+	// FailureDomain is the failure domain the machine will be created in.
+	// Must match a key in the FailureDomains map stored on the cluster object.
```
Comment (vincepri) [inline api/v1alpha3/machine_types.go]: Nice

**Pull request merged**